src/esw/drive/odrive_control.py

```diff
@@ -0,0 +1,656 @@
+        state data.
+    :param _vel_pub: A rospy Publisher object that is used for publishing
+        velocity data.
+    """
+
+    start_time: float
+    _current_lim: float
+    _id: str
+    _modrive: Modrive
+    _pair: str
+    _speed: Speed
+    _speed_lock: threading.Lock
+    _state: State
+    _state_pub: rospy.Publisher
+    _vel_pub: rospy.Publisher
+
+    def __init__(self, pair: str) -> None:
+        """Initializes the components, starting with a Disconnected State.
+        :param pair: A string that is front, middle, or back"""
+        self.start_time = t.process_time()
+        self._current_lim = rospy.get_param("/odrive/config/current_lim")
+        self._speed = Speed()
+        _odrive_ids = {
+            "front": rospy.get_param("/odrive/ids/front"),
+            "middle": rospy.get_param("/odrive/ids/middle"),
+            "back": rospy.get_param("/odrive/ids/back"),
+        }
+        self._pair = pair
+        self._id = _odrive_ids[self._pair]
+        self._speed_lock = threading.Lock()
+        self._state = DisconnectedState()
+        self._state_pub = rospy.Publisher("drive_state_data", DriveStateData, queue_size=1)
+        self._vel_pub = rospy.Publisher("drive_vel_data", DriveVelData, queue_size=1)
+
+    def change_axis_speed(self, axis: Axis, speed: float) -> None:
+        """Sets the self._speed to the requested speeds. The speeds must be in
+        the range [-1.0, 1.0].
+        :param axis: An Axis object that is either Axis.LEFT or Axis.RIGHT
+        :param speed: A float that has the requested speed of the axis.
+        """
+        assert -1.0 <= speed <= 1.0, "speed must be in range[-1.0, 1.0]"
+        self._speed_lock.acquire()
+        if axis == Axis.LEFT: 
+            self._speed = Speed(speed, self._speed.right)
+        else:
+            self._speed = Speed(self._speed.left, speed)
+        self._speed_lock.release()
+
+    def get_state_string(self) -> str:
+        """Returns the state of the ODriveBridge as a short string.
+        The form is of "[insert_odrive_state]State". e.g. state is ErrorState,
+        so ros_msg.state is Error
+        :returns: A string that is the state of the ODriveBridge.
+        """
+        state = str(self._state)
+        return state[: len(state) - len("State")]
+
+    def ros_publish_data_loop(self) -> None:
+        """Publishes velocity and current data continuously."""
+        while not rospy.is_shutdown():
+            self._publish_encoder_msg()
+
+    def watchdog_while_loop(self) -> None:
+        """Calls the update() function continuously and checks if comms has
+        been lost.
+        A flag is set to keep track fo the state for when we have comms with
+        the base station or not.
+        """
+        previously_lost_comms = True
+        while not rospy.is_shutdown():
+            watchdog = t.process_time() - self.start_time
+            lost_comms = watchdog > 1.0
```

Comment:
Does the watchdog value of 1.0 work ok when running this with all the odrives connected? Just curious if that was arbitrarily chosen and if it's the minimum without being too small.

Reply:
good question... we didn't exactly test with comms disconnected so we should test if the watchdog timer is good

Reply:
ok we talked this through in person. basically we were concerned that the odrives would stop working if the watchdog was too short, but this is fine.